server/src/main/java/org/candlepin/policy/js/entitlement/AbstractEntitlementRules.java

```diff
@@ -321,56 +326,82 @@ private void postUnbindVirtLimit(PoolHelper postHelper,
                 // -1.
-                List<Pool> pools = postHelper.lookupBySubscriptionId(pool
+                List<Pool> pools = poolManager.lookupBySubscriptionId(pool
                     .getSubscriptionId());
                 for (int idex = 0; idex < pools.size(); idex++) {
                     Pool derivedPool = pools.get(idex);
                     if (derivedPool.getAttributeValue("pool_derived") != null &&
                         derivedPool.getQuantity() == 0) {
-                        postHelper.setPoolQuantity(derivedPool, -1);
+                        poolManager.setPoolQuantity(derivedPool, -1);
                     }
                 }
             }
         }
     }
 
-    private void postBindVirtLimit(PoolHelper postHelper, Entitlement entitlement, Pool pool,
-        Consumer c, Map<String, String> attributes) {
+    private void postBindVirtLimit(PoolManager poolManager, Consumer c,
+            Map<String, Entitlement> entitlementMap, Map<String, Map<String, String>> attributeMaps,
+            List<Pool> subPoolsForStackIds) {
+        Set<String> stackIdsThathaveSubPools = new HashSet<String>();
+        if (CollectionUtils.isNotEmpty(subPoolsForStackIds)) {
+            for (Pool pool : subPoolsForStackIds) {
+                stackIdsThathaveSubPools.add(pool.getSourceStackId());
+            }
+        }
 
         log.debug("Running virt_limit post-bind.");
-        boolean hostLimited = attributes.containsKey("host_limited") &&
-            attributes.get("host_limited").equals("true");
-        if (!c.getType().isManifest() &&
-            !"true".equalsIgnoreCase(c.getFact("virt.is_guest")) &&
-            (config.getBoolean(ConfigProperties.STANDALONE) || hostLimited)) {
-            Product product = pool.getProduct();
-            String virtLimit = attributes.get("virt_limit");
 
-            String stackId = attributes.get("stacking_id");
-            boolean createSubPool = stackId == null ||
-                poolCurator.getSubPoolForStackId(c, stackId) == null;
-
-            if (createSubPool) {
-                log.debug("Creating a new sub-pool.");
-                try {
-                    int virtQuantity = Integer.parseInt(virtLimit);
-                    if (virtQuantity <= 0) {
-                        return;
+        boolean consumerFactExpression = !c.getType().isManifest() &&
+                !"true".equalsIgnoreCase(c.getFact("virt.is_guest"));
+
+        boolean isStandalone = config.getBoolean(ConfigProperties.STANDALONE);
+
+        List<Pool> createHostRestrictedPoolFor = new ArrayList<Pool>();
+        List<Entitlement> decrementHostedBonusPoolQuantityFor = new ArrayList<Entitlement>();
+
+        for (Entitlement entitlement : entitlementMap.values()) {
+            Pool pool = entitlement.getPool();
+            Map<String, String> attributes = attributeMaps.get(pool.getId());
+            boolean hostLimited = attributes.containsKey("host_limited") &&
```

Comment:
You could shorten this with `"true".equals(attributes.get("host_limited"))` which would set the boolean to false if the key were missing or if not set to "true"